File: python-Implementation/sieve_of_eratosthenes_prime.py

# Used to find all prime numbers up to a specified integer.
# It works by iteratively marking the multiples of each prime number starting from 2. 
# The numbers that remain unmarked at the end of the algorithm are the primes.

def sieve_of_eratosthenes(limit):
    primes = []
    prime_flags = [True] * (limit + 1)
    prime_flags[0] = prime_flags[1] = False  # 0 and 1 are not prime

    for p in range(2, int(limit**0.5) + 1):
        if prime_flags[p]:
            # Mark multiples of p as non-prime
            for i in range(p * p, limit + 1, p):
                prime_flags[i] = False

    for p in range(2, limit + 1):
        if prime_flags[p]:
            primes.append(p)

    return primes


def sieve(n):
    prime = [True for _ in range(n+1)]
    prime[0]=prime[1]=False
    p = 2
    while p*p <= n:
        if prime[p]:
            for i in range(p*p, n+1, p):
                prime[i] = False
        p += 1
    return [p for p in range(2, n) if prime[p]]

File: cpp-implementation/SieveOfEratosthenesPrime.cpp
#include <bits/stdc++.h>

using namespace std;

/*
Initialization:

Create a list of consecutive integers from 2 to 
n: (2, 3, 4, ..., n).
This list will be used to track numbers that are not eliminated; those that remain will be the prime numbers.
First Prime:

Start with the first prime number in the list, which is 2. This is the first prime number.
Eliminate Multiples:

Remove all multiples of 2 (excluding 2 itself) from the list. This means numbers: 4, 6, 8, 10, and so on, are removed.

The algorithm terminates when the square of the next prime exceeds 
n. All remaining unmarked numbers in the list are primes.
*/
// O(nlog log n)
vector<bool> sieveOfEratosthenes(int n) {
    vector<bool> isPrime(n + 1, true);
    isPrime[0] = isPrime[1] = false; // 0 and 1 are not prime numbers
    for (int i = 2; i * i <= n; i++) {
        if (isPrime[i]) {
            for (int j = i * i; j <= n; j += i) {
                isPrime[j] = false; // Mark all multiples of i as not prime
            }
        }
    }
    return isPrime;
}
